zebra-state/src/service/finalized_state/disk_db.rs

```diff
@@ -940,6 +943,111 @@ impl DiskDb {
+    /// used by the current db. If successful, it also deletes the db version file.
+    pub(crate) fn try_reusing_previous_db_after_major_upgrade(
+        restorable_db_versions: &[u64],
+        format_version_in_code: &Version,
+        config: &Config,
+        db_kind: impl AsRef<str>,
+        network: &Network,
+    ) {
+        if let Some(&major_db_ver) = restorable_db_versions
+            .iter()
+            .find(|v| **v == format_version_in_code.major)
+        {
+            let db_kind = db_kind.as_ref();
+
+            let old_path = config.db_path(db_kind, major_db_ver - 1, network);
+            let new_path = config.db_path(db_kind, major_db_ver, network);
+
+            let old_path = match fs::canonicalize(&old_path) {
+                Ok(canonicalized_old_path) => canonicalized_old_path,
+                Err(e) => {
+                    warn!("could not canonicalize {old_path:?}: {e}");
+                    return;
+                }
+            };
+
+            let cache_path = match fs::canonicalize(&config.cache_dir) {
+                Ok(canonicalized_cache_path) => canonicalized_cache_path,
+                Err(e) => {
+                    warn!("could not canonicalize {:?}: {e}", config.cache_dir);
+                    return;
+                }
+            };
+
+            // # Correctness
+            //
+            // Check that the path we're about to move is inside the cache directory.
+            //
+            // If the user has symlinked the state directory to a non-cache directory, we don't want
+            // to move it, because it might contain other files.
+            //
+            // We don't attempt to guard against malicious symlinks created by attackers
+            // (TOCTOU attacks). Zebra should not be run with elevated privileges.
+            if !old_path.starts_with(&cache_path) {
+                info!("skipped reusing previous state cache: state is outside cache directory");
+                return;
+            }
+
+            let opts = DiskDb::options();
+            let old_db_exists = DB::list_cf(&opts, &old_path).is_ok_and(|cf| !cf.is_empty());
+            let new_db_exists = DB::list_cf(&opts, &new_path).is_ok_and(|cf| !cf.is_empty());
+
+            if old_db_exists && !new_db_exists {
+                match fs::create_dir_all(&new_path) {
+                    Ok(()) => info!("created new directory for state cache at {new_path:?}"),
+                    Err(e) => {
+                        warn!(
+                            "could not create new directory for state cache at {new_path:?}: {e}"
+                        );
+                        return;
+                    }
+                };
+
+                match fs::rename(&old_path, &new_path) {
+                    Ok(()) => {
+                        info!("moved state cache from {old_path:?} to {new_path:?}");
+
+                        match fs::remove_file(new_path.join(DATABASE_FORMAT_VERSION_FILE_NAME)) {
```

Comment:
We should keep the version file for when multiple format upgrades need to be applied.

Reply:
The version file doesn't contain the major version, only the minor and patch versions. When we do a major version bump, we need to set the minor and patch versions to zero. If we transfer the db with its original version file containing non-zero versions, we'll end up with a malformed db version. Zebra recreates the version file if it's missing and sets the minor and patch versions to zero, so deleting the file is a simple means to reset the counters for minor and patch versions.